### DISCUSSION TITLE
how can I install forge on vsl?

### DISCUSSION THREAD

**Discussion by GenevaRoc** [Cyfrin/foundry-full-course-cu#841] [Q&A]
when i try to create a new project on foundry, using the forge init command i get this 
`Command 'forge' not found, but can be installed with:
sudo apt install snap`

**Comment by theirrationalone** [comment]
Please refer [this](https://github.com/Cyfrin/foundry-full-course-f23/discussions/801#discussioncomment-6960370)

**Reply by GenevaRoc** [reply]
Hi, sorry to bother again
I'm trying to run this command 'forge create SimpleStorage' it keeps throwing an error.
here is the error
....
`[⠒] Compiling...
No files changed, compilation skipped
Error: 
error sending request for url (http://localhost:8545/): error trying to connect: tcp connect error: Connection refused (os error 111)

Context:
- Error #0: error trying to connect: tcp connect error: Connection refused (os error 111)
- Error #1: tcp connect error: Connection refused (os error 111)
- Error #2: Connection refused (os error 111)`
.....
what can I do to fix it

**Reply by GenevaRoc** [reply]
nevermind, i figured it out

**Reply by theirrationalone** [reply]
Anvil was the suspect 😉

**Reply by GenevaRoc** [reply]
i had to reinstall foundry
deleted the file accidentally and didn't know, lol

**Comment by GenevaRoc** [comment]
hey!! i keep getting this error when i try to mintBasicNft 
`non-empty stderr args=["bash", "/home/genevaroc/.foundry/foundry-BasicNft-f23/lib/foundry-devops/src/get_recent_deployment.sh", "BasicNft", "11155111", "/home/genevaroc/.foundry/foundry-BasicNft-f23//broadcast"] stderr="/home/genevaroc/.foundry/foundry-BasicNft-f23/lib/foundry-devops/src/get_recent_deployment.sh: line 31: jq: command not found\n/home/genevaroc/.foundry/foundry-BasicNft-f23/lib/foundry-devops/src/get_recent_deployment.sh: line 33: jq: command not found\n/home/genevaroc/.foundry/foundry-BasicNft-f23/lib/foundry-devops/src/get_recent_deployment.sh: line 35: ((: i<: syntax error: operand expected (error token is \"<\")\n"
Traces:
  [639098] → new DevOpsTools@0x0e7494E39B8638CaA44581CB5A5dE4B973Da2459
    └─ ← 3192 bytes of code

  [228375] → new mintBasicNft@0x5b73C5498c1E3b4dbA84de0F1833c4a029d90519
    └─ ← 1030 bytes of code

  [59616] mintBasicNft::run() 
    ├─ [35626] DevOpsTools::e374cdf1(00000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000aa36a7000000000000000000000000000000000000000000000000000000000000000842617369634e6674000000000000000000000000000000000000000000000000) [delegatecall]
    │   ├─ [0] VM::ffi([pwd]) 
    │   │   └─ ← 0x2f686f6d652f67656e657661726f632f2e666f756e6472792f666f756e6472792d42617369634e66742d663233
    │   ├─ [0] VM::toString(11155111 [1.115e7]) [staticcall]
    │   │   └─ ← 0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000083131313535313131000000000000000000000000000000000000000000000000
    │   ├─ [0] VM::ffi([bash, /home/genevaroc/.foundry/foundry-BasicNft-f23/lib/foundry-devops/src/get_recent_deployment.sh, BasicNft, 11155111, /home/genevaroc/.foundry/foundry-BasicNft-f23//broadcast]) 
    │   │   └─ ← 0x0000000000000000000000000000000000000000
    │   ├─ [0] console::log(Return Data:) [staticcall]
    │   │   └─ ← ()
    │   ├─ [0] console::log(0x0000000000000000000000000000000000000000) [staticcall]
    │   │   └─ ← ()
    │   └─ ← 0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000144e6f20636f6e7472616374206465706c6f796564000000000000000000000000
    └─ ← "No contract deployed"



== Logs ==
  Return Data:
  0x0000000000000000000000000000000000000000
Error: 
No contract deployed`

**Reply by GenevaRoc** [reply]
here is my interactions code 

`contract mintBasicNft is Script {
    string public constant PUG =
        "ipfs://bafybeig37ioir76s7mg5oobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4/?filename=0-PUG.json";

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "BasicNft",
            block.chainid
        );
        mintNftOnContract(mostRecentlyDeployed);
    }

    function mintNftOnContract(address contractAddress) public {
        vm.startBroadcast();
        BasicNft(contractAddress).mintNft(PUG);
        vm.stopBroadcast();
    }
}
`

**Reply by theirrationalone** [reply]
Ahaaa `jq` is missing.... 

[Follow this](https://github.com/Cyfrin/foundry-full-course-f23/discussions/1008#discussioncomment-7330855) 

[And this too](https://github.com/Cyfrin/foundry-full-course-f23/discussions/1008#discussioncomment-7335709) 

;`)

**Reply by GenevaRoc** [reply]
just did. but now, it throws a different error of no contract deployed. do I redeploy my contract?

**Reply by theirrationalone** [reply]
Yes it could be, you should deploy it again with `--broadcast` subcommand and test again... You can verify your deployment by going to there on broadcast folder...

**Reply by GenevaRoc** [reply]
i still can't mint. it throws this error 
'[⠒] Compiling...
No files changed, compilation skipped
Error: 
Could not find target contract: script/Interactions.s.sol:MintBasicNFT
make: *** [Makefile:46: mint] Error 1'

My interactions file exist but i don't know why it says it find it

**Reply by GenevaRoc** [reply]
I have been able to fix the issue. Was a problem with spelling?

**Reply by theirrationalone** [reply]
Woww great 👍, glad to work with you ❤

**Reply by GenevaRoc** [reply]
hello! I keep having an error when i try to run this test `forge test --mt testConvertSvgToUri`

here is the error I'm having 
`Error (7576): Undeclared identifier.
  --> src/MoodNft.sol:40:14:
   |
40 |         if (!_isApprovedOrOwner(msg.sender, tokenId)) {
   |              ^^^^^^^^^^^^^^^^^^
`

and here is the code I'm getting the error in
`function flipMood(uint256 tokenId) public {
        //only want the Nft owner to be able to flip mood
        if (!_isApprovedOrOwner(msg.sender, tokenId)) {
            revert MoodNft__CantFlipMoodIfNotOwner();
        }
        if (s_tokenIdToMood[tokenId] == Mood.HAPPY) {
            s_tokenIdToMood[tokenId] = Mood.SAD;
        } else {
            s_tokenIdToMood[tokenId] = Mood.HAPPY;
        }
    }
`

**Reply by theirrationalone** [reply]
You might omitting to import its dependency.... Verify that....

**Reply by GenevaRoc** [reply]
the @openzeppelin file? I have that imported. Could it be that the name was updated from the parent contract?

**Reply by theirrationalone** [reply]
Yess it could be, you should go there and check if that function still exist Or openzeppelin has changed that utility function name Or derived some other new function for the same motive....

**Reply by GenevaRoc** [reply]
Still haven't been able to figure it out
